src/passes/analysis/verilog.cpp

```diff
@@ -1,93 +1,589 @@
+    } else if (auto bit_vector_value = dyn_cast<ConstBitVector>(value)) {
+        BitVector bit_vector = bit_vector_value->get();
+        return std::make_unique<vAST::NumericLiteral>(
+            bit_vector.hex_digits(), bit_vector.bitLength(), false, vAST::HEX);
+    } else if (auto string_value = dyn_cast<ConstString>(value)) {
+        return std::make_unique<vAST::String>(string_value->toString());
     }
-    WriteModuleToStream(module, os);
-  }
+    coreir_unreachable();
 }
 
-void Passes::Verilog::writeToFiles(const std::string& dir,
-                                   std::unique_ptr<std::string> product_file) {
-  std::vector<std::string> products;
-  for (auto module : vmods.vmods) {
-    if (vmods._inline && module->inlineable) continue;
-    const std::string filename = module->modname + ".v";
-    products.push_back(filename);
-    const std::string full_filename = dir + "/" + filename;
-    std::ofstream fout(full_filename);
-    ASSERT(fout.is_open(), "Cannot open file: " + full_filename);
-    WriteModuleToStream(module, fout);
-    fout.close();
-  }
-  // Write out the product list, if requested.
-  if (!product_file) return;
-  std::ofstream fout(*product_file);
-  ASSERT(fout.is_open(), "Cannot open file: " + *product_file);
-  for (const auto& product : products) {
-    fout << product << "\n";
-  }
-  fout.close();
+// Given a signal named `id` and a type `type`, wrap the signal name in a
+// `Vector` node if the signal is of type Array
+std::variant<std::unique_ptr<vAST::Identifier>, std::unique_ptr<vAST::Vector>>
+process_decl(std::unique_ptr<vAST::Identifier> id, Type *type) {
+    if (type->getKind() == Type::TK_Array) {
+        ArrayType *array_type = cast<ArrayType>(type);
+        ASSERT(array_type->getElemType()->isBaseType(),
+               "Expected Array of Bits");
+        return std::make_unique<vAST::Vector>(
+            std::move(id),
+            std::make_unique<vAST::NumericLiteral>(
+                toString(array_type->getLen() - 1)),
+            std::make_unique<vAST::NumericLiteral>("0"));
+    }
+
+    // unpack named types to get the raw type (so we can check that it's a
+    // bit), iteratively because perhaps you can name and named type?
+    // This is just a safety check for internal code, to improve performance,
+    // we could guard this assert logic behind a macro
+    while (type->getKind() == Type::TK_Named) {
+        type = cast<NamedType>(type)->getRaw();
+    }
+    ASSERT(type->isBaseType(), "Expected Bit, or Array of Bits");
+
+    // If it's not an Array type, just return the original identifier
+    return std::move(id);
+}
+
+// Given a map of instances, return a vector of containing declarations for all
+// the output ports of each instance.  We return a vector of a variant type for
+// compatibility with the module AST node constructor, even though we only ever
+// create Wire nodes.
+std::vector<std::variant<std::unique_ptr<vAST::StructuralStatement>,
```

Comment:
This templating is difficult to parse. I would prefer the arguments to variant to each be on their own line.